Introducing new updates for operand encoding generation

AngelDev06 opened the pull request:
The changes are the following:

- `getCSOperandEncoding` is now a variadic template (needed to combine the encodings of sub-operands to form the full complex operand)
- The above change was used to fix hopefully all post indexed ldr/str instructions (the issue was that LLVM treats base + offset as seperate operands while capstone as one, so I combined the encodings)
- Combines Rn & Rm operands for vld/vst instructions since capstone also treats them as one
- All of the indexes generate should hopefully be based on ISA now
- Made a new function specific for ARM encoding outputs to cover some exceptions (including ldr/str/vld/vst mentioned above)

Let me know if there are any issues

AngelDev06 committed:
Added handling for suboperands on encoding generation & now indexes are based on ISA

wargio commented on llvm/utils/TableGen/PrinterCapstone.cpp:
```diff
@@ -2926,7 +2979,10 @@ void printInsnOpMapEntry(CodeGenTarget const &Target,
 
-    std::string const &Encoding = getCSOperandEncoding(CGI, Rec, ArgName);
+    // Might wrap this check into a function if more arch's implement operand encoding.
+    std::string const &Encoding =
+        Target.getName() == "ARM" ? arm::getOperandEncoding(CGI, Rec, ArgName)
+                                  : getCSOperandEncoding(CGI, Rec, ArgName);
```
i guess having a switch might be more useful?

AngelDev06 replied:
Yeah strings need to be turned to enum flags in that case. I would use a constant unordered map for that purpose

Rot127 commented on llvm/lib/Target/ARM/ARMInstrThumb2.td:
```diff
@@ -1777,29 +1777,29 @@ def t2STRHT  : T2IstT<0b01, "strht", IIC_iStore_bh_i>;
 // ldrd / strd pre / post variants
 
-let mayLoad = 1, hasSideEffects = 0 in
+let mayLoad = 1, hasSideEffects = 0, IM = IndexModePre in
```
Please remove the additional `IM` variable. The post and prefix addressing should be determined by the operand types used.

E.g. here:
Post-Index: Uses `addr_offset_none, t2am_imm8s4_offset)`
Pre-Index: Uses `t2addrmode_imm8s4_pre)`

I know it is tempting to hard code this, but we run into several problems if we continue with defining flags like this:

- We need to do it for all instructions. Possibly missing some during updates. And auto-sync is meant to remove as many manual steps as possible. Checking the correct usage of `IM` variables with every update, is an additional manual step.
- We want to upstream the changes in `td` files to LLVM (because they miss some information). Such additions wouldn't be accepted, because it could be determined in another way or needs a change in the inheritance of instruction classes.

Rot127 replied:
Also see note below how to do this change

AngelDev06 replied:
I remember hardcoding it because some post indexed instructions couldn't be determined by the operand type as you mentioned. I can check again if you want

Rot127 replied:
Please do. You can check Capstones ARM code, for which operand types the `post_index` flag is set (see: `ARMMapping.c`).

AngelDev06 replied:
So I made it print all operand's names (which are marked as post indexed) and I got the following output:
![Screenshot 2023-08-06 200442](https://github.com/capstone-engine/llvm-capstone/assets/134562527/ebfaea2c-b7ad-48f7-8f62-0584004e03f4)
Since this obviously isn't reliable I will do what I said in my other comment. i.e. I will use `ARM_set_detail_op_mem_offset` to extend the encoding.
Hope this works as expected

Rot127 replied:
Sounds good! I hope I'll find time in the next weeks to add the encoding for PPC and AArch64 too.

Rot127 commented on llvm/utils/TableGen/PrinterCapstone.cpp:
```diff
@@ -2690,6 +2692,57 @@ std::string getCSOperandEncoding(CodeGenInstruction const *CGI,
+// this we fetch the full encoding that combines both operands in the second
+// iteration. The first one will return an empty encoding (which capstone won't
+// use so we are good)
+std::string getOperandEncoding(CodeGenInstruction const *CGI,
```
You can merge both `getOperandEncoding()` functions and remove the exceptions.

`arm::getOperandEncoding()` basically handles memory operands. But those are already handled in [`addComplexOperands()`](https://github.com/capstone-engine/llvm-capstone/blob/84848aa01f23d6498b66f5f23981da0b328ad83c/llvm/utils/TableGen/PrinterCapstone.cpp#L2907). There we already iterate over the Sub-Operands.

Complex operands (liek the memory operands) are DAGs, with an operand as root and multiple sub-operands as children. For memory operands the root is `addrmode_whatever::addr` and the children are usually `GPR::base` and `imm32::offsimm`).

You can call `getOperandEncoding()` within `addComplexOperand` and:
- get the encoding of the whole complex operand.
OR
- get the encoding for each sub-operand.

It depends which is present in the LLVMs instruction encoding.

The exceptions you handle below, are complex operands which have the sub-operands encoded. But nothing for the complex operand).

The really really annoying part is, that whoever implemented this, was really inconsistent with the encoding. Almost always they only save the bits of complex operands and say nothing about the sub-ops.

For example here (instruction: `MVE_VSTRB16`; td class: `MVE_VLDRSTR_cw`):

The `addr` operand has three components in the ISA:
```
addr:
  A-bit: 1bit
  base: 3bits
  offsimm: 7bits
```

but they encode it like this without any detail:

```
  field bits<32> Inst = { 1, 1, 1, 0, 1, 1, 0, 1, addr{11}, 0, 0, 0, 0, addr{10}, addr{8}, addr{7}, Qd{2}, Qd{1}, Qd{0}, 0, 1, 1, 1, 0, 1, addr{6}, addr{5}, addr{4}, addr{3}, addr{2}, addr{1}, addr{0} };
```

Which of those bits belong to which sub-operand is not clear at all.

To fix this properly, we would need to change almost all instruction classes. Which takes at least a day. I pushed an example [here](https://github.com/capstone-engine/llvm-capstone/commit/a174482759a2571aeb0c2729911905950396f97c).

If you have time to do this, it would be wonderful. Though if not, it is fine as well. Comes on the TODO list.

Also take a look at [this commit](https://github.com/capstone-engine/llvm-capstone/commit/d48f9bb9019904cade0e2ae8ec0c6ffb8d56b93b). It should help you to to move the `getOperandEncoding()` function into `addComplexOperand`.

AngelDev06 replied:
Ah there is an issue regarding that. For post indexed instructions the base and the offset are completely different operands. Not part of the same complex operand. 
So `addComplexOperand` doesn't add the encoding of both of them but only one of them at a time. 
And since capstone treats them as one operand (the memory operand) it overrides the encoding of base with the one of offset when mapping. 
So the user only gets to know the encoding of the offset. And yes this whole thing happens for post indexed instructions. 
Pre indexed are fine

AngelDev06 replied:
And yes btw I have added a call to `getCSOperandEncoding` on `addComplexOperand` which gets the encoding of the entire complex operand

AngelDev06 replied:
Also the exceptions I handle are exceptions because the .td files are a bit broken.
You see there are small thumb ldr-like instructions which have `Rn` & `Rm` (both being part of the memory operand) next to each other with the name on the encoding bits being just adr. 
So it's impossible for me to determine which of them is `Rn` and which `Rm`.
This way since they are next to each other in the encoding, `getCSOperandEncoding` treats both of them as one operand piece of the memory operand even though they should be seperate.
For example checkout [encoding T1 of thumb ldr register](https://developer.arm.com/documentation/ddi0406/c/Application-Level-Architecture/Instruction-Details/Alphabetical-list-of-instructions/LDR--register--Thumb-?lang=en) which has `Rn` at bit position 3 and `Rm` at bit position 6.

Without the exception this would include both pieces in one and say for example that the operand piece starts on 3 and has bit width 6.

As for the last exception this one is for pc relative thumb2 ldr which doesn't include the pc in the encoding when it should.

Rot127 replied:
> And since capstone treats them as one operand (the memory operand) it overrides the encoding of base with the one of offset when mapping.

In this case please check in Capstone, if anything might get overwritten and extend it.

Other targets might not have that much broken `td` files. And for them we should add a one to one mapping of operand to encoding bits.
Even if they are currently combined in the memory operand. For other targets or in the future they might be not.

> For example checkout [encoding T1 of thumb ldr register](https://developer.arm.com/documentation/ddi0406/c/Application-Level-Architecture/Instruction-Details/Alphabetical-list-of-instructions/LDR--register--Thumb-?lang=en) which has Rn at bit position 3 and Rm at bit position 6.

I see. Please fix those in the `td` files then. Since they are not that many I guess.
Hard coding them has the danger of adding wrong indices, when new instructions use this type but have different encoding.

> As for the last exception this one is for pc relative thumb2 ldr which doesn't include the pc in the encoding when it should.

Also for this one.

AngelDev06 replied:
> In this case please check in Capstone, if anything might get overwritten and extend it.

Yeah since the issue only happens for post index instructions, what I could do is modify the function that adds the post indexed memory operand on `ARMMapping.c`. iirc it's `ARM_add_mem_offset` which adds both parts of the memory operand. I could make it extend the encoding of base with the one of offset. (instead of overriding like it currently does) 


> I see. Please fix those in the `td` files then. Since they are not that many I guess.

Yeah I will try.

AngelDev06 commented:
@Rot127 alright so now post indexed instructions are handled by capstone. This generates the encoding for base & offset seperately and capstone is responsible for merging them. And yes I tested this out as well as pushed the commit with the changes on my other PR. 
I also added the exceptions on the `.td` files as you mentioned.

Rot127 commented on llvm/lib/Target/ARM/ARMInstrThumb2.td:
```diff
@@ -3912,26 +3912,24 @@ def t2TBH_JT : t2PseudoInst<(outs),
                     "tbb", "\t$addr", []>, Sched<[WriteBrTbl]> {
-  bits<4> Rn;
-  bits<4> Rm;
+  bits<8> addr;
   let Inst{31-20} = 0b111010001101;
-  let Inst{19-16} = Rn;
+  let Inst{19-16} = addr{7-4};
   let Inst{15-5} = 0b11110000000;
   let Inst{4} = 0; // B form
-  let Inst{3-0} = Rm;
```
You can leave these like it was (since this is more correct):
If you check in `addComplexOperand()` for the encoding of the names of the sub-operand (`Rn`, `Rm`) it will return it:

This was my result.
```
{ /* ARM_t2TBB (4290) - ARM_INS_TBB - tbb${p}   $addr */
{
  { CS_OP_MEM | CS_OP_REG, CS_AC_READ, { CS_DATA_TYPE_i32, CS_DATA_TYPE_LAST }, { 1, { 16 }, { 4 } } }, /* addr - Rn */
  { CS_OP_MEM | CS_OP_REG, CS_AC_READ, { CS_DATA_TYPE_i32, CS_DATA_TYPE_LAST }, { 1, { 0 }, { 4 } } }, /* addr - Rm */
...
```
And later combine the encodings to one in Capstone.

Used the example branch from above for this.

AngelDev06 replied:
Yeah I know it's `Rn` & `Rm` and this is also true in the ARM documentation. 

The thing is, this is the only exception in all of the ARM `td` files. 
The rest of the records that cover memory operands aren't descriptive enough and just put `addr` on the bits.
`tbb` & `tbh` are the only exceptions I could find that actually put the names of the suboperands in the instruction bits and therefore break the encoding generator

Rot127 replied:
That ARM is so broken in this regard is bad luck. But keep in mind, we will use it for other targets. And will fix the `td` files in the future. So a conceptually correct implementation is more important than one which fits just ARM.

AngelDev06 replied:
Yeah if you ask me you guys should include the suboperands of a memory operand in the instruction bits in each record.

I know this isn't going to be fun but at least it would make it easy to get the encoding for each suboperand individually.

Anyways I will try to find a way to do this differently for memory operands that actually have a dag with suboperands included and I will see if it breaks something regarding ARM.

AngelDev06 commented:
Fixed the issue regarding suboperands' names.
When tried to generate new tables (after of course I removed my changes on `tbb` & `tbh` records) there were no changes.
That means that it took care of the suboperands for those records.
Hopefully this should work for any auto-sync arch.
I also squashed all commits to keep the history clean.

Rot127 commented:
Thanks a lot! I'll cherry-pick it tomorrow or the day after into the AArch64 branch.

XVilka commented:
@Rot127 had been there any updates on this one?

Rot127 commented:
I wanted to integrate it directly into the AArch64 PR and make some changes here as well, because AArch64 needs them. But I judged the time effort incorrectly :(
Going to spend Thursday and Friday on the last ARM fixes and then add this one to AArch64.

Rot127 commented:
@AngelDev06 When you have time, please rebase on top of the latest auto-sync branch. Sorry it took forever. You know how it is...

AngelDev06 committed:
remove variable redefinition

AngelDev06 commented:
@Rot127 should be ok now

Rot127 reviewed (approved):
Ok, lets get this merged. If there is other stuff coming up we do it in another PR.
@AngelDev06 Thanks a lot for the contribution!
@wargio @XVilka If you want, take another look at it. And merge if you are ok with it.
